lib/cli/src/cli.rs

```diff
@@ -204,5 +233,244 @@ pub fn wasmer_main() {
+                        let command = sv.package.clone();
+                        if let Ok(o) = result {
+                            package_download_info = Some(o.clone());
+                            sp.close();
+                            sv.package = o.package;
+                            sv.version = Some(o.version);
+                            sv.command = Some(command);
+                            break;
+                        }
+                    }
+                }
+
+                if let Some(package) =
+                    wasmer_registry::get_local_package(&sv.package, sv.version.as_deref())
+                {
+                    let local_package_wasm_path = wasmer_registry::get_package_local_wasm_file(
+                        &package.registry,
+                        &package.name,
+                        &package.version,
+                    )
+                    .map_err(|e| anyhow!("{e}"))?;
+
+                    // Try finding the local package
+                    let mut args_without_package = args.clone();
+                    args_without_package.remove(1);
+
+                    let mut run_args = RunWithoutFile::try_parse_from(args_without_package.iter())?;
+                    run_args.command_name = sv.command.clone();
+                    return run_args
+                        .into_run_args(local_package_wasm_path, Some(package.manifest))
+                        .execute();
+                }
+
+                // else: local package not found
+                let sp =
+                    spinner::SpinnerBuilder::new(format!("Installing package {} ...", sv.package))
+                        .spinner(vec![
+                            "⣾", "⣽", "⣻", "⢿", "⡿", "⣟", "⣯", "⣷", " ", "⠁", "⠂", "⠄", "⡀", "⢀",
+                            "⠠", "⠐", "⠈",
+                        ])
+                        .start();
+
+                let v = sv.version.as_deref();
+                let result =
+                    wasmer_registry::install_package(&sv.package, v, package_download_info);
+                sp.close();
+                print!("\r\n");
+                match result {
+                    Ok((package, buf)) => {
+                        // Try auto-installing the remote package
+                        let mut args_without_package = args.clone();
+                        args_without_package.remove(1);
+
+                        let mut run_args =
+                            RunWithoutFile::try_parse_from(args_without_package.iter())?;
+                        run_args.command_name = sv.command.clone();
+
+                        return run_args
+                            .into_run_args(buf, Some(package.manifest))
+                            .execute();
+                    }
+                    Err(e) => {
+                        println!("{e}");
+                        return Ok(());
+                    }
+                }
+            }
+        }
```

Comment:
Yeah, I understand that the logic gets a lot harder because we're essentially trying to do half of `clap`'s job. 

My point is more that we've got a 180-line function that has 5-6 levels of indentation and which is performing several very different responsibilities.

I know you said you prefer to keep code together instead of splitting it out, but for someone that wasn't there when the code was written, long sprawling functions tend to make code really hard to read and maintain. They also tend to attract technical debt because you get into the *"this function is already pretty messy so it'll be fine to just add another if-statement"* mentality instead of continuously refactoring.

Reply:
I split it into multiple functions in https://github.com/wasmerio/wasmer/pull/3215/commits/471fd367209626455e5c9f48134d7cb5a0513570